timesketch/frontend/src/components/Explore/TimelineChip.vue

```diff
@@ -14,28 +14,320 @@ See the License for the specific language governing permissions and
+            </span>
+            <span>Delete</span>
+          </b-dropdown-item>
+        </b-dropdown>
+      </span>
+    </span>
   </div>
 </template>
 
 <script>
+import Vue from 'vue'
+import { Chrome } from 'vue-color'
+import _ from 'lodash'
+
+import TsAnalyzerHistory from '../Analyze/AnalyzerHistory'
+
+import EventBus from '../../main'
+
 export default {
-  props: ['timeline'],
+  components: {
+    'color-picker': Chrome,
+    TsAnalyzerHistory,
+  },
+  props: ['timeline', 'eventsCount', 'isSelected', 'isEmptyState'],
+  data() {
+    return {
+      initialColor: {},
+      newColor: '',
+      newTimelineName: '',
+      timelineStatus: '',
+      colorPickerActive: false,
+      showInfoModal: false,
+      showEditModal: false,
+      showAnalyzerModal: false,
+      isDarkTheme: false,
+    }
+  },
   computed: {
-    timelineColorStyle() {
-      let backgroundColor = this.timeline.color
+    meta() {
+      return this.$store.state.meta
+    },
+    datasourceErrors() {
+      return this.timeline.datasources.filter(datasource => datasource.error_message)
+    },
+    failedImportsMessage() {
+      return this.datasourceErrors.length + " failed imports"
+    },
+  },
+  methods: {
+    showColorPicker() {
+      this.$refs.colorPicker.click()
+    },
+    rename() {
+      this.showEditModal = false
+      this.$emit('save', this.timeline, this.newTimelineName)
+    },
+    remove() {
+      if (confirm('Delete the timeline?')) {
+        this.$emit('remove', this.timeline)
+      }
+    },
+    updateColor: _.debounce(function(color) {
+      this.newColor = color.hex
+      if (this.newColor.startsWith('#')) {
+        this.newColor = this.newColor.substring(1)
+      }
+      Vue.set(this.timeline, 'color', this.newColor)
+      this.$emit('save', this.timeline)
+    }, 300),
+    toggleTheme: function() {
+      this.isDarkTheme = !this.isDarkTheme
+    },
+    timelineColor(timeline) {
+      this.isDarkTheme = localStorage.theme === 'dark'
+      let backgroundColor = timeline.color
+      let textDecoration = 'none'
+      let opacity = '100%'
       if (!backgroundColor.startsWith('#')) {
         backgroundColor = '#' + backgroundColor
       }
+      // Grey out the index if it is not selected.
+      if (!this.isSelected) {
+        backgroundColor = '#d2d2d2'
+        textDecoration = 'line-through'
+        opacity = '50%'
+      }
+
+      if (this.isDarkTheme) {
+        return {
+          'background-color': backgroundColor,
+          'text-decoration': textDecoration,
+          opacity: opacity,
+          filter: 'grayscale(25%)',
+          color: '#333333',
+        }
+      }
       return {
         'background-color': backgroundColor,
+        'text-decoration': textDecoration,
+        opacity: opacity,
       }
     },
+    toggleTimeline: function(timeline) {
+      this.$emit('toggle', timeline)
+    },
+  },
+  mounted() {
+    // Hide color picker when clicked outside.
+    let self = this
```

Comment:
What's the reason we're assigning `this` to `self`? Can't we just do these operations on `this`?

Reply:
To be completely honest, I copied this function from a previous component, so I didn't inspect the purpose of this variable. At first glance, it looks completely unnecessary, let me check if it's still working fine without it.

Reply:
As suspected, it was there on purpose, as "this" becomes something else when the event is triggered compared to when the page is "mounted".

```
TimelineChip.vue?f907:291 Uncaught TypeError: Cannot read property 'contains' of undefined
    at eval (TimelineChip.vue?f907:291)
```

Reply:
I've added a comment though, to make it clear that this code is there for a good reason. Thanks for challenging it!

Reply:
My guess is that `this` is referencing the window object since that's where the function is defined? Ideally, you could add a method in the Vue object and call this method:

```
window.addEventListener('click', this.hideColorPicker(e))
```

(there's already a `showColorPicker`, I suspect for the same reason)